crypto/params.c

```diff
@@ -0,0 +1,401 @@
+{
+    return ossl_param_construct(key, OSSL_PARAM_REAL, buf, sizeof(double),
+                                rsize);
+}
+
+int OSSL_PARAM_get_double(const OSSL_PARAM *p, double *val)
+{
+#define CASE(type) \
+    case sizeof(type): \
+        *val = (double)(*(type *)p->buffer); \
+        return 1
+
+    if (val == NULL || p == NULL)
+        return 0;
+
+    switch (p->data_type) {
+    case OSSL_PARAM_REAL:
+        switch (p->buffer_size) {
+        CASE(double);
+        }
+        break;
+    case OSSL_PARAM_INTEGER:
+        switch (p->buffer_size) {
+        CASE(int32_t);
+        CASE(int64_t);
+        }
+        break;
+    case OSSL_PARAM_UNSIGNED_INTEGER:
+        switch (p->buffer_size) {
+        CASE(uint32_t);
+        CASE(uint64_t);
+        }
+    }
+    return 0;
+#undef CASE
+}
+
+int OSSL_PARAM_set_double(const OSSL_PARAM *p, double val)
+{
+#define CASE(type) \
+    case sizeof(type): \
+        *(type *)p->buffer = (type)val; \
+        SET_RETURN_SIZE(p, sizeof(type)); \
+        return 1
+
+    if (p == NULL)
+        return 0;
+
+    SET_RETURN_SIZE(p, sizeof(double));
+    switch (p->data_type) {
+    case OSSL_PARAM_REAL:
+        switch (p->buffer_size) {
+        CASE(double);
+        }
+        break;
+    case OSSL_PARAM_INTEGER:
+        switch (p->buffer_size) {
+        CASE(int32_t);
+        CASE(int64_t);
+        }
+        break;
+    case OSSL_PARAM_UNSIGNED_INTEGER:
+        switch (p->buffer_size) {
+        CASE(uint32_t);
+        CASE(uint64_t);
+        }
+        break;
+    }
+    return 0;    
+#undef CASE
+}
+
+/*******************************************************************
+* Track the allocated pointers using a sparse array.               *
+* This needs to be replaced with something better & more portable. *
+*******************************************************************/
+#include "internal/sparse_array.h"
+DEFINE_SPARSE_ARRAY_OF(void);
+static SPARSE_ARRAY_OF(void) *allocated_params = NULL;
+
+static void ossl_params_cleanup(void)
+{
+    ossl_sa_void_free(allocated_params);
+    allocated_params = NULL;
+}
+
+static CRYPTO_ONCE params_init_flag = CRYPTO_ONCE_STATIC_INIT;
+DEFINE_RUN_ONCE_STATIC(do_params_init)
+{
+    allocated_params = ossl_sa_void_new();
+    return allocated_params != NULL && OPENSSL_atexit(&ossl_params_cleanup);
+}
+
+static void ossl_param_register_alloc(void *string)
+{
+    if (!RUN_ONCE(&params_init_flag, do_params_init))
+        return;
+
+    /* Magic to go here */
+    ossl_sa_void_set(allocated_params, (size_t)string, string);
+}
+
+void OSSL_PARAM_string_free(void *string)
+{
+    if (!RUN_ONCE(&params_init_flag, do_params_init))
+        return;
+
+    /* Magic to go here */
+    if (ossl_sa_void_get(allocated_params, (size_t)string) == NULL)
+        return;
+    ossl_sa_void_set(allocated_params, (size_t)string, NULL);
+    free(string);
+}
+
+OSSL_PARAM OSSL_PARAM_construct_utf8_string(const char *key, double *buf,
```

Comment:
I suspect `double` is a copy'n'paste gaffe

Reply:
Yep :(